src/components/typography/Heading.tsx

```diff
@@ -0,0 +1,79 @@
+/*
+ * Echoes React
+ * Copyright (C) 2023-2024 SonarSource SA
+ * mailto:info AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+import styled from '@emotion/styled';
+import { PropsWithChildren } from 'react';
+
+export enum HeadingSize {
+  ExtraSmall = 'xsmall',
+  Small = 'small',
+  Medium = 'medium',
+  Large = 'large',
+  ExtraLarge = 'xlarge',
+}
+
+const bottomMarginByHeadingSize = (headingSize: HeadingSize) => {
+  switch (headingSize) {
+    case HeadingSize.Large:
+      return 'var(--echoes-dimension-space-200)';
+    case HeadingSize.ExtraLarge:
+      return 'var(--echoes-dimension-space-250)';
+    default:
+      return 'var(--echoes-dimension-space-100)';
+  }
+};
+
+type HeadingTag = 'h1' | 'h2' | 'h3' | 'h4' | 'h5';
+
+const defaultSizeByTag: Record<HeadingTag, HeadingSize> = {
+  h1: HeadingSize.ExtraLarge,
+  h2: HeadingSize.Large,
+  h3: HeadingSize.Medium,
+  h4: HeadingSize.Small,
+  h5: HeadingSize.ExtraSmall,
+};
+
+interface Props {
+  as: HeadingTag;
+  hasMarginBottom?: boolean;
+  size?: HeadingSize;
+}
+
+export function Heading({
+  as,
+  children,
+  hasMarginBottom = false,
+  size: sizeProp,
+}: Readonly<PropsWithChildren<Props>>) {
+  const size = sizeProp ?? defaultSizeByTag[as];
+
+  return <StyledHeading {...{ as, hasMarginBottom, size }}>{children}</StyledHeading>;
```

Comment:
why not pass the props like normal?

Reply:
it's just more concise

Reply:
mmmh so you would suggest that we switch to this notation more generally? I'd suggest bringing it to the weekly before we decide to introduce a new coding style on our own.

Reply:
I don't consider this "a new coding style", it's just more compact when the prop name and its value are spelled the same way. This isn't the first time this has been used either.

Reply:
I'm not so sure. I don't remember reviewing this pattern. If it was commonly used in our codebase I wouldn't be surprised seeing it here.

Reply:
we have a lot of `{...props}`, which is essentially the same thing, and a lot of objects where we don't have `foo: foo,` but just `foo,` - we even have a rule to make sure this is the case. So I'm not sure what's confusing, here?

Reply:
Ok sure, but I don't think it's nearly the same. We use `{...props}` because there is no other way.
So far, we have always written JSX the JSX way as much as possible, using a syntax that's close to HTML. We are all used to reading and understanding this syntax now and it's a widespread way of doing in the React/JSX community.

To me, this is a new coding style in which we aim for something more concise than our old plain JSX way. Applying this way of doing more generally will change a lot the way our components look (codewise).

It does impact all our devs, so is it something we really want to adopt? And why?

Reply:
I'm not a fan of it (i.e. no strong opinion, although I would probably not write it that way), but I think we can keep it as-is, and raise the topic in our weekly.
(this is in echoes, it's not really shared with anyone else)

Reply:
I've made this change:

```
-    <StyledHeading {...{ as, ref, size, ...props }} />
+    <StyledHeading as={as} ref={ref} size={size} {...props} />

```
is this OK for you guys?